### PR TITLE
[Relay/TOPI] Added 'offsets' and 'alignment' attributes to MATRIX_SET_DIAG.

### PR DESCRIPTION
**Pull request by jainris**
This is a follow-up PR of #6303.

* Added support for 'offsets' and 'alignment' attributes of MATRIX_SET_DIAG.
  (Similar to MATRIX_SET_DIAG V3 of TF)
* Added tests for 'offsets' and 'alignment' attributes of MATRIX_SET_DIAG.

### CODE REVIEW

**Comment by jainris** [comment]
cc @siju-samuel @mbaret @anijain2305 @u99127 @FrozenGene @tqchen

**Review by u99127 — APPROVED** [review]
Lgtm. Might need to run it through black for the python bits .

Ramana

**Comment by zhiics** [comment]
@siju-samuel @mbaret @anijain2305 @FrozenGene please help review. Thanks.

File: include/tvm/topi/transform.h [hunk 1]
```diff
@@ -1524,29 +1524,60 @@ inline Tensor sparse_to_dense(const Tensor& sparse_indices, const Array<Integer>
...
  * \return new tensor with given diagonal values.
  */
-inline Tensor matrix_set_diag(const Tensor& input, const Tensor& diagonal,
+inline Tensor matrix_set_diag(const Tensor& input, const Tensor& diagonal, int k1, int k2,
+                              bool super_diag_right_align, bool sub_diag_right_align,
                               const std::string name = "T_matrix_set_diag",
                               const std::string tag = kInjective) {
   size_t ndim = input->shape.size() - 1;
 
+  bool onlyOneDiagonal = k1 == k2;
```
Comment (mbaret) [inline include/tvm/topi/transform.h:1544]: only_one_diagonal

File: src/relay/op/tensor/transform.cc [hunk 2]
```diff
@@ -3131,36 +3133,62 @@ bool MatrixSetDiagRel(const Array<Type>& types, int num_inputs, const Attrs& att
   const auto* diagonal = types[1].as<TensorTypeNode>();
   CHECK(diagonal);
 
+  const auto param = attrs.as<MatrixSetDiagAttrs>();
+  CHECK_GE(param->k2, param->k1);
+
```
Comment (mbaret) [inline src/relay/op/tensor/transform.cc:3138]: Could you also add a check to make sure the k values are in a sane range for the input tensor size?

**Review by mbaret — CHANGES_REQUESTED** [review]

**Comment by mbaret** [comment]
ping @siju-samuel could you take a look at this as I believe you requested this functionality?

File: python/tvm/relay/op/transform.py [hunk 3]
```diff
@@ -1178,17 +1178,33 @@ def sparse_to_dense(sparse_indices, output_shape, sparse_values, default_value=0
...
 
     Parameters
     ----------
     data : relay.Expr
         Input Tensor.
+
     diagonal : relay.Expr
         Values to be filled in the diagonal.
 
+    k : int or tuple of int
```
Comment (siju-samuel) [inline python/tvm/relay/op/transform.py:1193]: add `Optional` for k and align

File: tests/python/relay/test_op_level10.py [hunk 4]
```diff
@@ -520,17 +518,18 @@ def _verify(input_shape, dtype):
...
                 out_relay = intrp.evaluate(func)(input_np, diagonal_np)
                 tvm.testing.assert_allclose(out_relay.asnumpy(), out_np)
 
-    _verify((2, 2), "float32")
-    _verify((4, 3, 3), "int32")
-    _verify((2, 3, 4), "float32")
+    _verify((2, 2), (2,), "float32")
+    _verify((4, 3, 3), (4, 3), "int32")
+    _verify((2, 3, 4), (2, 3), "float32", 1)
+    _verify((2, 3, 4), (2, 4, 3), "int32", (-1, 2), "LEFT_RIGHT")
```
Comment (siju-samuel) [inline tests/python/relay/test_op_level10.py:532]: add testcases for  "LEFT_LEFT" & "RIGHT_RIGHT"

**Review by siju-samuel — COMMENTED** [review]
LGTM, just some small suggestions

File: python/tvm/topi/transform.py [hunk 5]
```diff
@@ -806,17 +806,33 @@ def sparse_to_dense(sparse_indices, output_shape, sparse_values, default_value=0
...
 
     Parameters
     ----------
     data : relay.Expr
         Input Tensor.
+
     diagonal : relay.Expr
         Values to be filled in the diagonal.
 
+    k : int or tuple of int
```
Comment (anijain2305) [inline python/tvm/topi/transform.py:821]: Set optional as Samuel mentioned earlier

**Review by anijain2305 — APPROVED** [review]
LGTM. After samuel's comments are addressed, I can merge

**Review by mbaret — APPROVED** [review]
lgtm

**Comment by jainris** [comment]
@anijain2305 I believe this is ready to be merged.

**Review by anijain2305 — APPROVED** [review]

**Comment by anijain2305** [comment]
Thanks @jainris @mbaret @siju-samuel @u99127 This is merged!

**Pull request merged**